requirements.txt

```diff
@@ -1,10 +1,10 @@
-hepunits==1.1.*
-particle==0.9.*
-decaylanguage==0.6.*;python_version>="3.4"
-awkward==0.12.*
+hepunits==1.2.*
+particle==0.11.*
+decaylanguage==0.7.*
+awkward>=0.12.0,<1.0
```

Comment:
```suggestion
hepunits~=1.2.0
particle~=0.11.0
decaylanguage~=0.7.0
awkward~=0.12
```

Reply:
You don't need the `.0` on the end with that, right?

Reply:
Or is that how it decides "compatible releases"?

Reply:
Correct, the level (major, minor, patch) that you specify down to matters here. As an example:
```
$ docker run --rm -it python:3.8 /bin/bash
root@3984881cd2ed:/# pip install -q --upgrade pip setuptools wheel
root@3984881cd2ed:/# pip install -q "awkward~=0.12"
root@3984881cd2ed:/# pip list | grep awkward
awkward    0.13.0
root@3984881cd2ed:/# pip install -q "awkward~=0.12.0"
root@3984881cd2ed:/# pip list | grep awkward
awkward    0.12.22
```

Reply:
Thanks for the thoughts. Interesting alternative. Sounds good.

I conclude from the above that one does want the 0 in, since what is meant to be specified e.g. in `awkward>=0.12.0,<1.0` is "take any version above and including 0.12.0, and less than 1.0, so anything 0.* above 0.12.0.

Reply:
I will accept all suggestions and will make the tiny update.

Reply:
> I conclude from the above that one does want the 0 in

No sorry the opposite.

Here
```
awkward>=0.12.0,<1.0
```
the final `0` is unnecessary as it doesn't convey any more information than `v0.12`. However

```
awkward~=0.12.0
```

will pin it to the `v0.12.X` range, but the original intent seems to be just to restrict it to the `v0.X` range. So the lack of `0` was intentional.

Reply:
Yesterday I installed docker on my Windows laptop and have just given your commands above a try. Yes, I see what you meant and now finally understand your point :-). Thanks very much.

Reply:
Sorry for adding potential confusion :) but I think how 695fbe13c256b42d497a196c3d18e6892f71adcc has it now is right. :+1: That being
```diff
- awkward>=0.12.0,<1.0
+ awkward~=0.12
```

Reply:
The goal is IMO the following: have a compatible set of packages of the form x.y.* with the dependencies the best possible. So, if we pin `uproot` to a certain x.y then `uproot-methods` and `awkward-array` can be the best/higher version that is still compatible. Right? Then the minor .0 is only necessary for `uproot`. Am I getting this right? Your 2 most recent comments contradict this.